src/core/primitives.rs

```diff
@@ -451,3 +438,53 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Attempts opening the Covercrypt encapsulation using the given USK. Returns
+/// the encapsulated key and associated rights upon success, otherwise returns `None`.
+pub fn full_decaps(
+    encapsulation: &XEnc,
+    msk: &MasterSecretKey,
+) -> Result<Vec<(Right, Secret<SHARED_SECRET_LENGTH>)>, Error> {
+    // A = ⊙ _i (α_i. c_i)
+    let A = encapsulation.c.first().unwrap() * &(&msk.tsk.s / &msk.tsk.tracers.front().unwrap().0);
+    let mut rights_list: Vec<(Right, Secret<SHARED_SECRET_LENGTH>)> = Vec::new();
+    for enc in &encapsulation.encapsulations {
+        for (right, key) in msk.secrets.iter() {
+            for k in key {
+                let list: Vec<Right> = rights_list.clone().into_iter().map(|r| r.0).collect();
+                if list.contains(right) {
```

Comment:
Searching in a vector is `O(n)` which is highly inefficient. Just break upon pushing a new secret.